### PR TITLE
KAFKA-14857: Fix some MetadataLoader bugs

### PR DESCRIPTION
**Pull request by cmccabe**
The MetadataLoader is not supposed to publish metadata updates until we have loaded up to the high water mark. Previously, this logic was broken, and we published updates immediately. This PR fixes that and adds a junit test.

Another issue is that the MetadataLoader previously assumed that we would periodically get callbacks from the Raft layer even if nothing had happened. We relied on this to install new publishers in a timely fashion, for example. However, in older MetadataVersions that don't include NoOpRecord, this is not a safe assumption.

### CODE REVIEW

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 1]
```diff
@@ -212,38 +216,67 @@ private MetadataLoader(
         this.uninitializedPublishers = new LinkedHashMap<>();
         this.publishers = new LinkedHashMap<>();
         this.image = MetadataImage.EMPTY;
-        this.eventQueue = new KafkaEventQueue(time, logContext,
+        this.eventQueue = new KafkaEventQueue(Time.SYSTEM, logContext,
```
Comment (dengziming) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:219]: Is this change related to test?

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 2]
```diff
@@ -212,38 +216,67 @@ private MetadataLoader(
...
-                    offset + ", but the high water mark is " + highWaterMark.getAsLong());
+            log.info("{}: The loader is still catching up because we have loaded up to offset " +
+                    offset + ", but the high water mark is {}", where, highWaterMark.getAsLong());
             return true;
         }
-        log.info("The loader finished catch up to the current high water mark of " +
-                highWaterMark.getAsLong());
+        log.info("{}: The loader finished catching up to the current high water mark of {}",
+                where, highWaterMark.getAsLong());
         catchingUp = true;
```
Comment (dengziming) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: should we change this to catchingUp= false?

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 3]
```diff
@@ -256,23 +289,26 @@ private void maybeInitializeNewPublishers() {
...
-                faultHandler.handleFault("Unhandled error publishing the initial metadata " +
-                        "image from snapshot at offset " + image.highestOffsetAndEpoch().offset() +
-                        " with publisher " + publisher.name(), e);
+                faultHandler.handleFault("Unhandled error initializing " + publisher.name() +
+                        " with a snapshot at offset " + image.highestOffsetAndEpoch().offset(), e);
             }
         }
     }
 
+    String uninitializedPublisherNames() {
```
Comment (dengziming) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: nit: we could make this method private.

**Review by dengziming — COMMENTED** [review]
Thank you for this PR @cmccabe , I left some comments.

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 4]
```diff
@@ -212,38 +216,67 @@ private MetadataLoader(
...
-            log.trace("There are no uninitialized publishers to initialize.");
+            log.debug("InitializeNewPublishers: nothing to do.");
+            return;
+        }
+        if (stillNeedToCatchUp("initializeNewPublishers", image.highestOffsetAndEpoch().offset())) {
+            // Reschedule the initialization for later.
+            log.debug("InitializeNewPublishers: unable to initialize new publisher(s) {} " +
+                            "because we are still catching up with quorum metadata. Rescheduling.",
+                    uninitializedPublisherNames());
+            scheduleInitializeNewPublishers(TimeUnit.MILLISECONDS.toNanos(1));
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: The high-watermark accessor is reading the HW from RaftClient. Is there a case where we would be waiting on fetches from RaftClient before we had sufficient logs to be caught up? I'm wondering if 1ms is too tight for cases where we're waiting on some network requests.

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 5]
```diff
@@ -256,23 +289,26 @@ private void maybeInitializeNewPublishers() {
                 image.provenance(),
                 time.nanoseconds() - startNs);
         for (Iterator<MetadataPublisher> iter = uninitializedPublishers.values().iterator();
-                iter.hasNext(); ) {
+                 iter.hasNext(); ) {
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: nit: is this the right indentation?

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 2 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: yes. good catch

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 5 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: fixed

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 4 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: Fair point. I'll make it 100 ms but also schedule it immediately after handleCommit / handleSnapshot if needed to avoid bloating junit test time

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 1 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:219]: Using mock time doesn't really work with deferred event queue stuff at the moment. I do want to fix this. But right now, it would make `MockTime` unusable to pass it to `KafkaEventQueue`.

I will add a time accessor to `KafkaEventQueue` so it is at least clear when we're using the event queue time

**Review by mumrah — APPROVED** [review]
LGTM

**Review by dengziming — APPROVED** [review]
LGTM.
